modules/pricer/src/main/java/com/opengamma/strata/pricer/calibration/MarketDataFxRateProvider.java

```diff
@@ -33,7 +35,19 @@ public double fxRate(Currency baseCurrency, Currency counterCurrency) {
     if (baseCurrency.equals(counterCurrency)) {
       return 1;
     }
-    FxRate rate = marketData.getValue(FxRateKey.of(baseCurrency, counterCurrency));
-    return rate.fxRate(baseCurrency, counterCurrency);
+    // Try direct pair
+    Optional<FxRate> rate = marketData.findValue(FxRateKey.of(baseCurrency, counterCurrency));
+    if (rate.isPresent()) {
+      return rate.get().fxRate(baseCurrency, counterCurrency);
+    }
+    // Try triangulation on first currency
+    Currency triangularCcy = baseCurrency.getTriangulationCurrency();
+    Optional<FxRate> rate1 = marketData.findValue(FxRateKey.of(baseCurrency, triangularCcy));
+    Optional<FxRate> rate2 = marketData.findValue(FxRateKey.of(triangularCcy, counterCurrency));
+    if(rate1.isPresent() && rate2.isPresent()) {
+      return rate1.get().crossRate(rate2.get()).fxRate(baseCurrency, counterCurrency);
+    }
+    throw new IllegalArgumentException("No market data available for pair " + baseCurrency + "/" + counterCurrency);
```

Comment:
Before throwing an exception should we also try:

* counter currency triangulation currency
* triangulating twice if there are two triangulation currencies